rust/loqui_client/src/connection_handler.rs

```diff
@@ -0,0 +1,395 @@
+            };
+            match await!(reader.next()) {
+                Some(Ok(UpgradeFrame::Response)) => Ok(writer.reunite(reader)?.into_inner()),
+                Some(Ok(frame)) => Err(LoquiError::InvalidUpgradeFrame { frame }.into()),
+                Some(Err(e)) => Err(e),
+                None => Err(LoquiError::TcpStreamClosed.into()),
+            }
+        }
+    }
+
+    fn handshake(&mut self, mut reader_writer: ReaderWriter) -> Self::HandshakeFuture {
+        async move {
+            let hello = Self::make_hello();
+            reader_writer = match await!(reader_writer.write(hello)) {
+                Ok(read_writer) => read_writer,
+                Err(e) => return Err((e.into(), None)),
+            };
+
+            match await!(reader_writer.reader.next()) {
+                Some(Ok(frame)) => match Self::handle_handshake_frame(frame) {
+                    Ok(ready) => Ok((ready, reader_writer)),
+                    Err(e) => Err((e, Some(reader_writer))),
+                },
+                Some(Err(e)) => Err((e, Some(reader_writer))),
+                None => Err((LoquiError::TcpStreamClosed.into(), Some(reader_writer))),
+            }
+        }
+    }
+
+    fn handle_frame(
+        &mut self,
+        frame: DelegatedFrame,
+        encoder: Arc<Box<Encoder<Encoded = F::Encoded, Decoded = F::Decoded>>>,
+    ) -> Option<Self::HandleFrameFuture> {
+        match frame {
+            DelegatedFrame::Response(response) => {
+                self.handle_response(response, encoder);
+                None
+            }
+            DelegatedFrame::Error(error) => {
+                self.handle_error(error);
+                None
+            }
+            DelegatedFrame::Push(_) | DelegatedFrame::Request(_) => Some(
+                async move {
+                    Err((
+                        LoquiError::InvalidOpcode {
+                            actual: Request::OPCODE,
+                            expected: None,
+                        }
+                        .into(),
+                        0,
+                    ))
+                },
+            ),
+        }
+    }
+
+    fn handle_internal_event(
+        &mut self,
+        event: InternalEvent<F::Encoded, F::Decoded>,
+        id_sequence: &mut IdSequence,
+        encoder: Arc<Box<Encoder<Encoded = F::Encoded, Decoded = F::Decoded>>>,
+    ) -> Option<LoquiFrame> {
+        // Forward Request and Push events to the connection so it can send them to the server.
+        match event {
+            InternalEvent::Request { payload, waiter } => {
+                let sequence_id = id_sequence.next();
+                self.handle_request(payload, sequence_id, waiter, encoder)
+            }
+            InternalEvent::Push { payload } => self.handle_push(payload, encoder),
+        }
+    }
+
+    fn handle_ping(&mut self) {
+        // Use to sweep dead waiters.
+        let now = Instant::now();
```

Comment:
It is very much a hack, heh. We need a way for the connection handler to be able to sweep occasionally.

We aren't notifying the client here because it should have already stopped waiting. Also, the drop of the waiter_tx here stops the client from waiting any longer.

Reply:
I renamed it to `on_ping_received`.